### PR TITLE
Ice pushnotification data collection

### PR DESCRIPTION
**Pull request by toiletspider**
### References Issues:
[API] Push notifications data tracking

### Description:
Add end points to count sent and opened push notifications.
Also added the broadcast id to the push notification extra data field.

### Migrations
 * No Migrations

### Environment Variables
 * No change

### CODE REVIEW

File: db/src/models/broadcasts.rs [hunk 1]
```diff
@@ -76,9 +80,33 @@ impl Broadcast {
...
 
+    pub fn increment_open_count(
+        id: Uuid,
+        connection: &PgConnection,
+    ) -> Result<Broadcast, DatabaseError> {
+        let broadcast = Broadcast::find(id, connection)?;
+        diesel::update(&broadcast)
+            .set(broadcasts::dsl::opened_quantity.eq(broadcast.opened_quantity + 1))
+            .get_result(connection)
+            .to_db_error(ErrorCode::UpdateError, "Unable to update broadcast")
```
Comment (mikethetike) [inline db/src/models/broadcasts.rs]:
```suggestion
            .to_db_error(ErrorCode::UpdateError, "Unable to update open count on broadcast")
```

File: api/src/controllers/broadcasts.rs [hunk 2]
```diff
@@ -112,3 +112,19 @@ pub fn delete(
...
+    Ok(HttpResponse::Ok().finish())
+}
+
+pub fn tracking_count_sent(
+    (conn, path, _user): (Connection, Path<PathParameters>, User),
+) -> Result<HttpResponse, BigNeonError> {
+    let connection = conn.get();
+    Broadcast::increment_open_count(path.id, connection)?;
+    Ok(HttpResponse::Ok().finish())
+}
```
Comment (Krakaw) [inline api/src/controllers/broadcasts.rs]: @toiletspider we need to store the sent_count in the `BroadcastPushNotificationExecutor::perform_job` using ` audience.length()`

image

File: api/src/controllers/broadcasts.rs [hunk 3]
```diff
@@ -112,3 +112,19 @@ pub fn delete(
...
+    let connection = conn.get();
+    Broadcast::increment_sent_count(path.id, connection)?;
+    Ok(HttpResponse::Ok().finish())
+}
+
+pub fn tracking_count_sent(
+    (conn, path, _user): (Connection, Path<PathParameters>, User),
+) -> Result<HttpResponse, BigNeonError> {
+    let connection = conn.get();
+    Broadcast::increment_open_count(path.id, connection)?;
```
Comment (Krakaw) [inline api/src/controllers/broadcasts.rs]: This is the wrong function.

File: api/src/controllers/broadcasts.rs [hunk 4]
```diff
@@ -112,3 +112,19 @@ pub fn delete(
     let broadcast = broadcast.cancel(connection)?;
     Ok(HttpResponse::Ok().json(broadcast))
 }
+
+pub fn tracking_count_open(
+    (conn, path, _user): (Connection, Path<PathParameters>, User),
+) -> Result<HttpResponse, BigNeonError> {
+    let connection = conn.get();
+    Broadcast::increment_sent_count(path.id, connection)?;
```
Comment (Krakaw) [inline api/src/controllers/broadcasts.rs]: This is the wrong function.

File: api/src/routing.rs [hunk 5]
```diff
@@ -39,6 +39,12 @@ pub fn routes(app: &mut CorsBuilder<AppState>) -> App<AppState> {
         r.method(Method::PUT).with(broadcasts::update);
         r.method(Method::DELETE).with(broadcasts::delete);
     })
+    .resource("/broadcasts/tracking_count_open/{id}", |r| {
+        r.method(Method::GET).with(broadcasts::tracking_count_open);
+    })
+    .resource("/broadcasts/tracking_count_sent/{id}", |r| {
+        r.method(Method::GET).with(broadcasts::tracking_count_sent);
+    })
```
Comment (Krakaw) [inline api/src/routing.rs]: We don't need this endpoint

File: api/src/routing.rs [hunk 6]
```diff
@@ -39,6 +39,12 @@ pub fn routes(app: &mut CorsBuilder<AppState>) -> App<AppState> {
         r.method(Method::PUT).with(broadcasts::update);
         r.method(Method::DELETE).with(broadcasts::delete);
     })
+    .resource("/broadcasts/tracking_count_open/{id}", |r| {
```
Comment (Krakaw) [inline api/src/routing.rs]: We are using the standard REST pattern.
```
.resource("/broadcasts/{id}/tracking_count_open", |r| {
```

**Pull request merged**